### ISSUE TITLE
Error : cannot inherit non-frozen dataclass from a frozen one

### ISSUE THREAD

**Issue by EtienneFerrandi** [huggingface/trl#1013]
I try to fine-tune Llama 2 and when I launch the training with :

```
trainer = SFTTrainer(
    model=model,
    train_dataset=dataset,
    peft_config=peft_config,
    dataset_text_field="text",
    max_seq_length=max_seq_length,
    tokenizer=tokenizer,
    args=training_arguments,
    packing=packing,
)
```

I have this error :

```
[/usr/lib/python3.10/dataclasses.py](https://localhost:8080/#) in _process_class(cls, init, repr, eq, order, unsafe_hash, frozen, match_args, kw_only, slots)
    983         # Raise an exception if any of our bases are frozen, but we're not.
    984         if any_frozen_base and not frozen:
--> 985             raise TypeError('cannot inherit non-frozen dataclass from a '
    986                             'frozen one')
    987 

TypeError: cannot inherit non-frozen dataclass from a frozen one.
```

Which parameter of the SFTTrainer could be linked with this error ?

**Comment by lvwerra** [comment]
Maybe if you post the full stack trace it will be clearer. What version of TRL are you using? Can you try to update?

**Comment by EtienneFerrandi** [comment]
Here is the full stack trace :

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-41-9af62b74a122>](https://localhost:8080/#) in <cell line: 31>()
     29 )
     30 
---> 31 trainer.train() #Pour un nouveau fine-tuning.
     32 #trainer.train(resume_from_checkpoint=True) #Pour prolonger un fine-tuning existant.
     33 trainer.model.save_pretrained(output_dir)

18 frames
[/usr/lib/python3.10/dataclasses.py](https://localhost:8080/#) in _process_class(cls, init, repr, eq, order, unsafe_hash, frozen, match_args, kw_only, slots)
    983         # Raise an exception if any of our bases are frozen, but we're not.
    984         if any_frozen_base and not frozen:
--> 985             raise TypeError('cannot inherit non-frozen dataclass from a '
    986                             'frozen one')
    987 

TypeError: cannot inherit non-frozen dataclass from a frozen one
```

And trl version :

`!pip show trl`

```
Name: trl
Version: 0.7.4
Summary: A Pytorch implementation of Proximal Policy Optimization for transfomer language models.
Home-page: https://github.com/huggingface/trl
Author: Leandro von Werra
Author-email: [leandro.vonwerra@gmail.com](mailto:leandro.vonwerra@gmail.com)
License: Apache 2.0
Location: /usr/local/lib/python3.10/dist-packages
Requires: accelerate, datasets, numpy, torch, transformers, tyro
Required-by:
```

**Comment by EtienneFerrandi** [comment]
My peft_config : 

```
peft_config = LoraConfig(
    lora_alpha=lora_alpha,
    lora_dropout=lora_dropout,
    r=lora_r,
    inference_mode=False,
    task_type="CAUSAL_LM"
)
```

**Comment by vwxyzjn** [comment]
@EtienneFerrandi thanks for reporting the issue. Could you provide a minimal repro? I tried running with the following and encountered no issues. It mirrors your setup of using py3.10

```py
from datasets import load_dataset
from trl import SFTTrainer

dataset = load_dataset("imdb", split="train")

trainer = SFTTrainer(
    "facebook/opt-350m",
    train_dataset=dataset,
    dataset_text_field="text",
    max_seq_length=512,
)
trainer.train()
```
```
costa@ip-26-0-155-73:/fsx/costa/trl-test$ python test.py 
/fsx/costa/trl-test/.venv/lib/python3.10/site-packages/trl/trainer/ppo_config.py:141: UserWarning: The `optimize_cuda_cache` arguement will be deprecated soon, please use `optimize_device_cache` instead.
  warnings.warn(
/fsx/costa/trl-test/.venv/lib/python3.10/site-packages/trl/trainer/sft_trainer.py:145: UserWarning: You passed a model_id to the SFTTrainer. This will automatically create an `AutoModelForCausalLM` or a `PeftModel` (if you passed a `peft_config`) for you.
  warnings.warn(
Map: 100%|████████████████████████████████████████| 25000/25000 [00:04<00:00, 5450.08 examples/s]
/fsx/costa/trl-test/.venv/lib/python3.10/site-packages/torch/cuda/__init__.py:138: UserWarning: CUDA initialization: The NVIDIA driver on your system is too old (found version 11060). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org to install a PyTorch version that has been compiled with your version of the CUDA driver. (Triggered internally at ../c10/cuda/CUDAFunctions.cpp:108.)
  return torch._C._cuda_getDeviceCount() > 0
  0%|                                                                   | 0/9375 [00:00<?, ?it/s]You're using a GPT2TokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
  0%|                                                        | 4/9375 [00:59<38:14:56, 14.69s/it]
```

**Comment by vwxyzjn** [comment]
To check with PEFT:

```
from datasets import load_dataset
from trl import SFTTrainer
from peft import LoraConfig
dataset = load_dataset("imdb", split="train")

peft_config = LoraConfig(
    r=64,
    lora_alpha=16,
    lora_dropout=0.3,
    inference_mode=False,
    bias="none",
    task_type="CAUSAL_LM",
)
trainer = SFTTrainer(
    "facebook/opt-350m",
    train_dataset=dataset,
    dataset_text_field="text",
    max_seq_length=512,
    peft_config=peft_config,
)
trainer.train()
```

```
(trl-test-py3.10) costa@ip-26-0-155-73:/fsx/costa/trl-test$ python test.py 
/fsx/costa/trl-test/.venv/lib/python3.10/site-packages/trl/trainer/ppo_config.py:141: UserWarning: The `optimize_cuda_cache` arguement will be deprecated soon, please use `optimize_device_cache` instead.
  warnings.warn(
/fsx/costa/trl-test/.venv/lib/python3.10/site-packages/trl/trainer/sft_trainer.py:145: UserWarning: You passed a model_id to the SFTTrainer. This will automatically create an `AutoModelForCausalLM` or a `PeftModel` (if you passed a `peft_config`) for you.
  warnings.warn(
  0%|                                                            | 0/9375 [00:00<?, ?it/s]You're using a GPT2TokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
  0%|                                                  | 9/9375 [00:05<1:21:12,  1.92it/s]
```

**Comment by EtienneFerrandi** [comment]
Yes, thanks for your answer.

I load the model like this :

```
compute_dtype = getattr(torch, "float16")

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype = "float16",
    bnb_4bit_use_double_quant=False,
)

if compute_dtype == torch.float16 and use_4bit:
    major, _ = torch.cuda.get_device_capability()
    if major >= 8:
        print("=" * 80)
        print("Your GPU supports bfloat16, you can accelerate training with the argument --bf16")
        print("=" * 80)

model = AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map=device_map,
    quantization_config=bnb_config,
)

model.config.use_cache = False
model.config.pretraining_tp = 1
```

Then I changed the peft_config (added target_modules) and it raised the error.

```
peft_config = LoraConfig(
    lora_alpha=lora_alpha,
    lora_dropout=lora_dropout,
    r=lora_r,
    inference_mode=False,
    task_type="CAUSAL_LM",
    target_modules = ["q_proj", "v_proj"] 
)
```

But I have a new one :
`trainer.train()`
```
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/linear.py in forward(self, input)
    112 
    113     def forward(self, input: Tensor) -> Tensor:
--> 114         return F.linear(input, self.weight, self.bias)
    115 
    116     def extra_repr(self) -> str:

RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument mat2 in method wrapper_CUDA_mm)
```

**Comment by vwxyzjn** [comment]
This seems like a separate issue. Can you repro the non-froezen dataclass issue?
